File: src/log_migration/asset_manifest.py
import hashlib
import json
import mimetypes
from dataclasses import dataclass
from pathlib import Path
from urllib.parse import urlsplit, urlunsplit

from .normalize import NormalizationResult


_IMAGE_HOSTS = {"gyazo.com", "i.gyazo.com"}


@dataclass(frozen=True)
class AssetManifestEntry:
    id: str
    url: str
    kind: str
    source_post_ids: tuple[str, ...]

    def as_dict(self) -> dict[str, object]:
        return {
            "id": self.id,
            "url": self.url,
            "kind": self.kind,
            "source_post_ids": list(self.source_post_ids),
        }


def canonicalize_url(url: str) -> str:
    parsed = urlsplit(url.strip())
    if parsed.scheme.lower() not in {"http", "https"} or not parsed.netloc:
        raise ValueError(f"unsupported external URL: {url!r}")
    return urlunsplit(
        (
            parsed.scheme.lower(),
            parsed.netloc.lower(),
            parsed.path,
            parsed.query,
            "",
        )
    )


def stable_url_asset_id(url: str) -> str:
    canonical_url = canonicalize_url(url)
    digest = hashlib.sha256(canonical_url.encode("utf-8")).hexdigest()[:16]
    return f"asset_{digest}"


def classify_url(url: str) -> str:
    parsed = urlsplit(url)
    host = (parsed.hostname or "").lower()
    if host in _IMAGE_HOSTS:
        return "image"

    mime_type, _ = mimetypes.guess_type(parsed.path)
    if mime_type is None:
        return "url"
    if mime_type.startswith("image/"):
        return "image"
    if mime_type.startswith("audio/"):
        return "audio"
    if mime_type.startswith("video/"):
        return "video"
    return "url"


def build_asset_manifest(
    normalized: NormalizationResult,
) -> tuple[AssetManifestEntry, ...]:
    sources: dict[str, set[str]] = {}
    kinds: dict[str, str] = {}
    for post in normalized.posts:
        for raw_url in post.external_urls:
            try:
                canonical_url = canonicalize_url(raw_url)
            except ValueError:
                continue
            sources.setdefault(canonical_url, set()).add(post.id)
            kinds.setdefault(canonical_url, classify_url(canonical_url))

    entries = tuple(
        AssetManifestEntry(
            id=stable_url_asset_id(url),
            url=url,
            kind=kinds[url],
            source_post_ids=tuple(sorted(sources[url])),
        )
        for url in sorted(sources)
    )
    return tuple(sorted(entries, key=lambda entry: entry.id))


def write_asset_manifest(
    output_dir: Path,
    normalized: NormalizationResult,
) -> Path:
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    path = output_dir / "asset-manifest.json"
    payload = {
        "assets": [entry.as_dict() for entry in build_asset_manifest(normalized)]
    }
    path.write_text(
        json.dumps(payload, ensure_ascii=False, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )
    return path

File: lib/weblog_migration/asset_manifest.rb
# frozen_string_literal: true

require "digest"
require "json"
require "uri"
require "pathname"
require "fileutils"

module WeblogMigration
  module AssetManifest
    IMAGE_HOSTS = %w[gyazo.com i.gyazo.com].freeze
    MIME_TYPES = {
      ".jpg" => "image/jpeg", ".jpeg" => "image/jpeg", ".png" => "image/png", ".gif" => "image/gif",
      ".webp" => "image/webp", ".svg" => "image/svg+xml", ".mp3" => "audio/mpeg", ".wav" => "audio/wav",
      ".m4a" => "audio/mp4", ".ogg" => "audio/ogg", ".mp4" => "video/mp4", ".webm" => "video/webm",
      ".mov" => "video/quicktime",
    }.freeze

    module_function

    def canonicalize_url(url)
      raw_url = url.to_s.strip
      match = raw_url.match(%r{\A(https?)://([^/?#]+)([^?#]*)(?:\?([^#]*))?(?:#.*)?\z}i)
      raise ArgumentError, "unsupported external URL: #{url.inspect}" if match.nil?

      scheme = match[1].downcase
      authority = match[2].downcase
      path = match[3]
      query = match[4]
      "#{scheme}://#{authority}#{path}#{query.nil? ? "" : "?#{query}"}"
    rescue ArgumentError
      raise ArgumentError, "unsupported external URL: #{url.inspect}"
    end

    def stable_url_asset_id(url)
      "asset_#{Digest::SHA256.hexdigest(canonicalize_url(url))[0, 16]}"
    end

    def stable_asset_id(source_path)
      "asset_#{Digest::SHA256.hexdigest(source_path.to_s)[0, 16]}"
    end

    def classify_url(url)
      parsed = HTTP.request_uri(url)
      return "image" if IMAGE_HOSTS.include?(parsed.host&.downcase)

      mime_type = MIME_TYPES[File.extname(parsed.path).downcase]
      return "url" if mime_type.nil?
      return "image" if mime_type.start_with?("image/")
      return "audio" if mime_type.start_with?("audio/")
      return "video" if mime_type.start_with?("video/")

      "url"
    rescue URI::InvalidURIError
      "url"
    end

    def build_asset_manifest(normalized)
      sources = Hash.new { |hash, key| hash[key] = [] }
      kinds = {}
      normalized.posts.each do |post|
        post.external_urls.each do |raw_url|
          begin
            canonical_url = canonicalize_url(raw_url)
          rescue ArgumentError
            next
          end
          sources[canonical_url] << post.id unless sources[canonical_url].include?(post.id)
          kinds[canonical_url] ||= classify_url(canonical_url)
        end
      end
      sources.keys.sort.map do |url|
        AssetManifestEntry.new(
          id: stable_url_asset_id(url),
          url:,
          kind: kinds.fetch(url),
          source_post_ids: sources.fetch(url).sort
        )
      end.sort_by(&:id)
    end

    def write_asset_manifest(output_dir, normalized)
      output_dir = Pathname(output_dir)
      FileUtils.mkdir_p(output_dir)
      path = output_dir.join("asset-manifest.json")
      payload = { "assets" => build_asset_manifest(normalized).map(&:to_h) }
      path.write(JSON.pretty_generate(payload) + "\n", encoding: "UTF-8")
      path
    end
  end
end
